### PR TITLE
A little backend cleanup

### LINKED ISSUE #1067: Design and clean up error types linking
### Proposal

We have a lot of error types and reasons them to occur.

Need to somehow unite them by providing soft api for converting from lower level to upper one. 

Remove duplicates (`TerminationReason` and `TerminationReasonKind`) and remove from everywhere Option<trap_reason>, cause it should be always known for now (opt is legit for initial state and optional storing while running backend, any further transformations should be concrete)

Remove `description` field from exts, etc.

Waiting for any other thoughts on how to implement.

### PR DESCRIPTION
**Pull request by ark0f**
Resolves #1067.

### CODE REVIEW

File: core-backend/sandbox/src/env.rs [hunk 1]
```diff
@@ -46,16 +46,18 @@ use sp_sandbox::{
...
     #[display(fmt = "Unable to set module memory data")]
     SetModuleMemoryData,
     #[display(fmt = "Unable to save static pages initial data")]
     SaveStaticPagesInitialData,
-    #[display(fmt = "Failed to create env memory")]
-    CreateEnvMemory,
+    #[display(fmt = "Failed to create env memory: {:?}", _0)]
+    CreateEnvMemory(sp_sandbox::Error),
+    #[display(fmt = "No trap explanation")]
+    NoTrapExplanation,
```
Comment (breathx) [inline core-backend/sandbox/src/env.rs:60]: There should be no cases without trap explanation. Our backend field keeps zero only for cases of `unreachable` section in wasm. This should be checked somehow and research if there are similar-behaviour reasons

File: core-backend/sandbox/src/env.rs [hunk 2]
```diff
@@ -73,7 +75,6 @@ pub(crate) struct Runtime<E: Ext> {
     pub ext: ExtCarrier<E>,
     pub memory: MemoryWrap,
     pub trap: Option<FuncError<E::Error>>,
```
Comment (breathx) [inline core-backend/sandbox/src/env.rs:77]: how do you feel to keep here not the Option with error, but the "actual TerminationReason"? It starts with `Success` and changes to other variants in appropriate cases? could we do something like that?

File: core-backend/sandbox/src/funcs.rs [hunk 3]
```diff
@@ -92,14 +92,10 @@ pub enum FuncError<E> {
     DebugString(FromUtf8Error),
     #[display(fmt = "`gr_error` expects error occurred earlier")]
     SyscallErrorExpected,
+    #[display(fmt = "Terminated: {:?}", _0)]
+    Terminated(TerminationReason),
     #[display(fmt = "`gr_exit` has been called")]
     Exit,
```
Comment (breathx) [inline core-backend/sandbox/src/funcs.rs:98]: Why doesn't `Exit` case included in `Terminated`?

File: core-backend/wasmtime/Cargo.toml [hunk 4]
```diff
@@ -11,6 +11,7 @@ gear-core-errors = { path = "../../core-errors", features = ["codec"] }
 gear-backend-common = { path = "../common" }
 
 wasmtime = { version = "0.35.1", default-features = false, features = ["parallel-compilation", "cranelift"] }
+anyhow = "1.0.57"
```
Comment (breathx) [inline core-backend/wasmtime/Cargo.toml:14]: should we disable default features?

Comment (breathx) [reply core-backend/wasmtime/Cargo.toml:14]: there is the `std` one

File: core-backend/wasmtime/src/env.rs [hunk 5]
```diff
@@ -50,20 +49,22 @@ pub struct StoreData<E: Ext> {
...
     #[display(fmt = "Unable to set module memory data")]
     SetModuleMemoryData,
     #[display(fmt = "Unable to save static pages initial data")]
     SaveStaticPagesInitialData,
-    #[display(fmt = "Failed to create env memory")]
-    CreateEnvMemory,
+    #[display(fmt = "Failed to create env memory: {}", _0)]
+    CreateEnvMemory(anyhow::Error),
+    #[display(fmt = "No trap explanation")]
+    NoTrapExplanation,
```
Comment (breathx) [inline core-backend/wasmtime/src/env.rs:67]: same as in sandbox

File: core-backend/wasmtime/src/funcs.rs [hunk 6]
```diff
@@ -829,12 +829,9 @@ where
...
         let func = move |mut caller: Caller<'_, StoreData<E>>| -> Result<(), Trap> {
-            caller.data_mut().termination_reason = Some(TerminationReason::Trap {
-                explanation: Some(TrapExplanation::Other(
-                    "Unable to call a forbidden function".into(),
-                )),
-                description: None,
-            });
+            caller.data_mut().termination_reason = Some(TerminationReason::Trap(
+                TrapExplanation::Other("Unable to call a forbidden function".into()),
+            ));
```
Comment (breathx) [inline core-backend/wasmtime/src/funcs.rs:834]: Probably better to provide `TerminationReason::Trap(TrapExplanation::ForbiddenFunction)`?


By the way, for `TrapExplanation` `Cow` seems good, to don't allocate until we need to do something with the inner str

**Review by breathx — APPROVED** [review]
Looks great 👍 
I really want you to do some research to detect wasm native variants to break the execution, which we should detect to avoid cases with "Unknown error". They all should be transparent!

**Review by shamilsan — APPROVED** [review]
Going in the right direction 👍

File: pallets/gear/src/manager.rs [hunk 7]
```diff
@@ -258,15 +258,12 @@ where
...
-                log::trace!("Dispatch outcome trap: {:?}", message_id);
-
-                if let Some(reason) = trap {
-                    log::info!(
-                        "🪤 Program {} terminated with a trap: {}",
-                        program_id.into_origin(),
-                        reason
-                    );
-                };
+                log::trace!(
```
[the remark this answers is not in the document]
Comment (ark0f) [reply pallets/gear/src/manager.rs]: Oops, missed the fact there are different levels. Fixed

**Comment by breathx** [comment]
> 

Seems that first two cases covered by our platform logic
Can't the third case be detected on the instantiating time?
And the fourth seems to be alone. To this is not actually unknown reason, that's exactly `WasmRuntimeTrap` / unreachable section always

File: core-backend/sandbox/src/funcs.rs [hunk 8]
```diff
@@ -92,23 +93,27 @@ pub enum FuncError<E> {
...
         match self {
             Self::Core(err) => Some(err),
             _ => None,
         }
     }
+
+    pub fn into_termination_reason(self) -> TerminationReason {
+        match self {
+            Self::Terminated(reason) => reason,
+            err => TerminationReason::Trap(TrapExplanation::Other(err.to_string())),
```
Comment (breathx) [inline core-backend/sandbox/src/funcs.rs:114]: As I said previously, can't we pass here not string, but the cow? For future encoding, we don't need to allocate always.

File: core-backend/sandbox/src/env.rs [hunk 9]
```diff
@@ -72,8 +72,7 @@ pub struct SandboxEnvironment<E: Ext + IntoExtInfo> {
 pub(crate) struct Runtime<E: Ext> {
     pub ext: ExtCarrier<E>,
     pub memory: MemoryWrap,
-    pub trap: Option<FuncError<E::Error>>,
-    pub termination_reason: Option<TerminationReasonKind>,
+    pub trap: FuncError<E::Error>,
```
Comment (breathx) [inline core-backend/sandbox/src/env.rs:75]: that's not correct naming so, if it may be success termination.

**Review by breathx — APPROVED** [review]

File: core-backend/sandbox/src/funcs.rs (hunk 8 above)
Comment (breathx) [reply core-backend/sandbox/src/funcs.rs:114]: The lifecycle of errors would be
created -> thrown to upper layer -> (optionally) printed in `log::debug` -> encoded via `(E: Encode) trap.encode()` -> sent as bytes after encode.

`&'static str` encodes the same as `String`, but need to check it if I'm wrong, so no need in allocating string always .

Comment (ark0f) [reply core-backend/sandbox/src/funcs.rs:114]: I wondering do we actually need it now as `Other` variant is always built with `String` and is it premature optimisations or not

File: core-backend/wasmtime/Cargo.toml (hunk 4 above)
Comment (breathx) [reply core-backend/wasmtime/Cargo.toml:14]: https://github.com/gear-tech/gear/blob/3009d0e2fe986abf11511975e12056d1d0b6f1ee/core-backend/wasmtime/src/lib.rs#L21
Isn't it?

File: core-backend/sandbox/src/funcs.rs (hunk 8 above)
Comment (breathx) [reply core-backend/sandbox/src/funcs.rs:114]: Ok, lets left it as it is atm

**Pull request merged**